File: engine/io/effect_model_loader.py
"""Load and query Phase 2.5 tactical effect model artifacts."""

from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path
from typing import Dict, List


DEFAULT_EFFECT_MODEL_PATH = Path("compiled/tactical_effect_models_v1.json")


@lru_cache(maxsize=4)
def load_effect_model(path: str) -> Dict[str, object]:
    model_path = Path(path)
    return json.loads(model_path.read_text(encoding="utf-8"))


def lookup_hazard_source(
    hazard_id: str,
    source_name: str,
    source_type: str = "trigger_action",
    model_path: str | Path = DEFAULT_EFFECT_MODEL_PATH,
) -> Dict[str, object]:
    model = load_effect_model(str(model_path))
    entries = model.get("hazards", {}).get("entries", [])
    for hazard in entries:
        if hazard.get("hazard_id") != hazard_id:
            continue
        for source in hazard.get("sources", []):
            if source.get("source_type") == source_type and source.get("source_name") == source_name:
                return {
                    "hazard_id": hazard_id,
                    "hazard_name": hazard.get("hazard_name"),
                    "source_type": source_type,
                    "source_name": source_name,
                    "effects": source.get("effects", []),
                    "raw_text": source.get("raw_text"),
                }
    raise KeyError(f"hazard source not found: hazard_id={hazard_id} source_type={source_type} source_name={source_name}")

File: old/src/io/effectModelLoader.ts
/**
 * Load and query Phase 2.5 tactical effect model artifacts.
 * Mirrors engine/io/effect_model_loader.py
 *
 * In the browser this fetches JSON via fetch() or uses bundled data.
 * The default path is the compiled effect models artifact.
 */

const DEFAULT_EFFECT_MODEL_PATH = "data/rules/effect_models.json";

// Module-level cache
const modelCache = new Map<string, Record<string, unknown>>();

export async function loadEffectModel(
  path = DEFAULT_EFFECT_MODEL_PATH,
): Promise<Record<string, unknown>> {
  if (modelCache.has(path)) {
    return modelCache.get(path)!;
  }
  const response = await fetch(path);
  if (!response.ok) {
    throw new Error(`Failed to load effect model from ${path}: ${response.statusText}`);
  }
  const data = (await response.json()) as Record<string, unknown>;
  modelCache.set(path, data);
  return data;
}

// Synchronous variant for use in reducer (requires pre-loading)
let _preloadedModel: Record<string, unknown> | null = null;

export function setPreloadedEffectModel(model: Record<string, unknown>): void {
  _preloadedModel = model;
}

export function lookupHazardSource(
  hazardId: string,
  sourceName: string,
  sourceType = "trigger_action",
  modelPath?: string,
): Record<string, unknown> {
  const model = _preloadedModel;
  if (!model) {
    throw new Error(
      `Effect model not preloaded. Call loadEffectModel() and setPreloadedEffectModel() first. path=${modelPath ?? DEFAULT_EFFECT_MODEL_PATH}`,
    );
  }
  const hazards = model["hazards"] as Record<string, unknown> | undefined;
  const entries =
    (hazards?.["entries"] as Array<Record<string, unknown>>) ?? [];

  for (const hazard of entries) {
    if (hazard["hazard_id"] !== hazardId) continue;
    const sources =
      (hazard["sources"] as Array<Record<string, unknown>>) ?? [];
    for (const source of sources) {
      if (source["source_type"] === sourceType && source["source_name"] === sourceName) {
        return {
          hazard_id: hazardId,
          hazard_name: hazard["hazard_name"],
          source_type: sourceType,
          source_name: sourceName,
          effects: source["effects"] ?? [],
          raw_text: source["raw_text"],
        };
      }
    }
  }
  throw new Error(
    `hazard source not found: hazard_id=${hazardId} source_type=${sourceType} source_name=${sourceName}`,
  );
}
